File: Tau.KeyVault.GoClient/proto.go
package keyvault

import (
	"fmt"
	"io"
)

// ─────────────────────────────────────────────────────────
//  Minimal protobuf encoder/decoder for wire compatibility
//  with the server's protobuf-net [ProtoContract] types.
//
//  protobuf-net uses proto3 wire format. We implement just
//  enough of the protobuf wire format to encode/decode the
//  message types used by the API, avoiding a heavy codegen
//  dependency. Field numbers match [ProtoMember(n)].
// ─────────────────────────────────────────────────────────

// Wire types
const (
	wireVarint  = 0
	wireFixed64 = 1
	wireBytes   = 2
	wireFixed32 = 5
)

// ── Encoder ──────────────────────────────────────────────

type pbEncoder struct {
	buf []byte
}

func (e *pbEncoder) writeTag(field int, wt int) {
	e.writeVarint(uint64(field<<3 | wt))
}

func (e *pbEncoder) writeVarint(v uint64) {
	for v >= 0x80 {
		e.buf = append(e.buf, byte(v)|0x80)
		v >>= 7
	}
	e.buf = append(e.buf, byte(v))
}

func (e *pbEncoder) writeString(field int, s string) {
	if s == "" {
		return
	}
	e.writeTag(field, wireBytes)
	e.writeVarint(uint64(len(s)))
	e.buf = append(e.buf, s...)
}

func (e *pbEncoder) writeBool(field int, v bool) {
	if !v {
		return
	}
	e.writeTag(field, wireVarint)
	e.buf = append(e.buf, 1)
}

func (e *pbEncoder) writeInt32(field int, v int) {
	if v == 0 {
		return
	}
	e.writeTag(field, wireVarint)
	e.writeVarint(uint64(v))
}

func (e *pbEncoder) writeBytes(field int, b []byte) {
	if len(b) == 0 {
		return
	}
	e.writeTag(field, wireBytes)
	e.writeVarint(uint64(len(b)))
	e.buf = append(e.buf, b...)
}

func (e *pbEncoder) bytes() []byte { return e.buf }

// ── Decoder ──────────────────────────────────────────────

type pbDecoder struct {
	data []byte
	pos  int
}

func newDecoder(data []byte) *pbDecoder {
	return &pbDecoder{data: data}
}

func (d *pbDecoder) done() bool { return d.pos >= len(d.data) }

func (d *pbDecoder) readVarint() (uint64, error) {
	var v uint64
	var shift uint
	for {
		if d.pos >= len(d.data) {
			return 0, io.ErrUnexpectedEOF
		}
		b := d.data[d.pos]
		d.pos++
		v |= uint64(b&0x7F) << shift
		if b < 0x80 {
			return v, nil
		}
		shift += 7
		if shift >= 64 {
			return 0, fmt.Errorf("varint overflow")
		}
	}
}

func (d *pbDecoder) readTag() (field int, wt int, err error) {
	v, err := d.readVarint()
	if err != nil {
		return 0, 0, err
	}
	return int(v >> 3), int(v & 7), nil
}

func (d *pbDecoder) readBytes() ([]byte, error) {
	length, err := d.readVarint()
	if err != nil {
		return nil, err
	}
	if d.pos+int(length) > len(d.data) {
		return nil, io.ErrUnexpectedEOF
	}
	b := d.data[d.pos : d.pos+int(length)]
	d.pos += int(length)
	return b, nil
}

func (d *pbDecoder) readString() (string, error) {
	b, err := d.readBytes()
	return string(b), err
}

func (d *pbDecoder) readBool() (bool, error) {
	v, err := d.readVarint()
	return v != 0, err
}

func (d *pbDecoder) readInt() (int, error) {
	v, err := d.readVarint()
	return int(v), err
}

func (d *pbDecoder) skipField(wt int) error {
	switch wt {
	case wireVarint:
		_, err := d.readVarint()
		return err
	case wireFixed64:
		if d.pos+8 > len(d.data) {
			return io.ErrUnexpectedEOF
		}
		d.pos += 8
		return nil
	case wireBytes:
		_, err := d.readBytes()
		return err
	case wireFixed32:
		if d.pos+4 > len(d.data) {
			return io.ErrUnexpectedEOF
		}
		d.pos += 4
		return nil
	default:
		return fmt.Errorf("unknown wire type %d", wt)
	}
}

// ── Encode functions ─────────────────────────────────────

func encodeUpsertRequest(r *upsertRequest) []byte {
	e := &pbEncoder{}
	e.writeString(1, r.Key)
	e.writeString(2, r.Value)
	e.writeString(3, r.Environment)
	e.writeString(4, r.DataType)
	e.writeBool(5, r.IsSensitive)
	return e.bytes()
}

func encodeRenameRequest(r *renameRequest) []byte {
	e := &pbEncoder{}
	e.writeString(1, r.NewName)
	return e.bytes()
}

func encodeImportKeyItem(k *ImportKeyItem) []byte {
	e := &pbEncoder{}
	e.writeString(1, k.Key)
	e.writeString(2, k.Value)
	e.writeString(3, k.DataType)
	e.writeBool(4, k.IsSensitive)
	return e.bytes()
}

func encodeImportRequest(r *ImportRequest) []byte {
	e := &pbEncoder{}
	e.writeString(1, r.Environment)
	e.writeString(2, r.Mode)
	for i := range r.Keys {
		e.writeBytes(3, encodeImportKeyItem(&r.Keys[i]))
	}
	return e.bytes()
}

// ── Decode functions ─────────────────────────────────────

func decodeKeyEntryResponse(data []byte) (*KeyEntryResponse, error) {
	d := newDecoder(data)
	r := &KeyEntryResponse{}
	for !d.done() {
		field, wt, err := d.readTag()
		if err != nil {
			return nil, err
		}
		switch field {
		case 1:
			r.Key, err = d.readString()
		case 2:
			r.Value, err = d.readString()
		case 3:
			r.Environment, err = d.readString()
		case 4:
			r.DataType, err = d.readString()
		case 5:
			r.IsSensitive, err = d.readBool()
		case 6:
			// bcl.DateTime — skip (we prefer JSON for datetime)
			err = d.skipField(wt)
		default:
			err = d.skipField(wt)
		}
		if err != nil {
			return nil, err
		}
	}
	return r, nil
}

func decodeKeyEntryListResponse(data []byte) (*KeyEntryListResponse, error) {
	d := newDecoder(data)
	r := &KeyEntryListResponse{}
	for !d.done() {
		field, wt, err := d.readTag()
		if err != nil {
			return nil, err
		}
		switch field {
		case 1:
			b, err2 := d.readBytes()
			if err2 != nil {
				return nil, err2
			}
			item, err2 := decodeKeyEntryResponse(b)
			if err2 != nil {
				return nil, err2
			}
			r.Items = append(r.Items, *item)
		default:
			err = d.skipField(wt)
		}
		if err != nil {
			return nil, err
		}
	}
	return r, nil
}

func decodeEnvironmentListResponse(data []byte) (*EnvironmentListResponse, error) {
	d := newDecoder(data)
	r := &EnvironmentListResponse{}
	for !d.done() {
		field, _, err := d.readTag()
		if err != nil {
			return nil, err
		}
		switch field {
		case 1:
			s, err2 := d.readString()
			if err2 != nil {
				return nil, err2
			}
			r.Environments = append(r.Environments, s)
		default:
			// skip
		}
	}
	return r, nil
}

func decodeDeleteEnvironmentResponse(data []byte) (*DeleteEnvironmentResponse, error) {
	d := newDecoder(data)
	r := &DeleteEnvironmentResponse{}
	for !d.done() {
		field, wt, err := d.readTag()
		if err != nil {
			return nil, err
		}
		switch field {
		case 1:
			r.Message, err = d.readString()
		case 2:
			r.DeletedKeys, err = d.readInt()
		default:
			err = d.skipField(wt)
		}
		if err != nil {
			return nil, err
		}
	}
	return r, nil
}

func decodeRenameEnvironmentResponse(data []byte) (*RenameEnvironmentResponse, error) {
	d := newDecoder(data)
	r := &RenameEnvironmentResponse{}
	for !d.done() {
		field, wt, err := d.readTag()
		if err != nil {
			return nil, err
		}
		switch field {
		case 1:
			r.Message, err = d.readString()
		case 2:
			r.UpdatedKeys, err = d.readInt()
		default:
			err = d.skipField(wt)
		}
		if err != nil {
			return nil, err
		}
	}
	return r, nil
}

func decodeExportKeyItemResponse(data []byte) (*ExportKeyItemResponse, error) {
	d := newDecoder(data)
	r := &ExportKeyItemResponse{}
	for !d.done() {
		field, wt, err := d.readTag()
		if err != nil {
			return nil, err
		}
		switch field {
		case 1:
			r.Key, err = d.readString()
		case 2:
			r.Value, err = d.readString()
		case 3:
			r.DataType, err = d.readString()
		case 4:
			r.IsSensitive, err = d.readBool()
		default:
			err = d.skipField(wt)
		}
		if err != nil {
			return nil, err
		}
	}
	return r, nil
}

func decodeExportPayloadResponse(data []byte) (*ExportPayloadResponse, error) {
	d := newDecoder(data)
	r := &ExportPayloadResponse{}
	for !d.done() {
		field, wt, err := d.readTag()
		if err != nil {
			return nil, err
		}
		switch field {
		case 1:
			r.Version, err = d.readString()
		case 2:
			// bcl.DateTime — skip
			err = d.skipField(wt)
		case 3:
			r.Environment, err = d.readString()
		case 4:
			r.KeyCount, err = d.readInt()
		case 5:
			b, err2 := d.readBytes()
			if err2 != nil {
				return nil, err2
			}
			item, err2 := decodeExportKeyItemResponse(b)
			if err2 != nil {
				return nil, err2
			}
			r.Keys = append(r.Keys, *item)
		default:
			err = d.skipField(wt)
		}
		if err != nil {
			return nil, err
		}
	}
	return r, nil
}

func decodeImportResultResponse(data []byte) (*ImportResultResponse, error) {
	d := newDecoder(data)
	r := &ImportResultResponse{}
	for !d.done() {
		field, wt, err := d.readTag()
		if err != nil {
			return nil, err
		}
		switch field {
		case 1:
			r.Imported, err = d.readInt()
		case 2:
			r.Skipped, err = d.readInt()
		case 3:
			r.Message, err = d.readString()
		default:
			err = d.skipField(wt)
		}
		if err != nil {
			return nil, err
		}
	}
	return r, nil
}

File: Tau.KeyVault.PyClient/tau_keyvault/proto.py
"""
Protobuf codec — runtime-defined message descriptors that are wire-compatible
with the server's [ProtoContract] / [ProtoMember] annotations (protobuf-net).

protobuf-net uses field numbers from [ProtoMember(n)] and proto3 semantics.
DateTime is serialized as protobuf-net's bcl.DateTime surrogate (sint64 value +
int32 scale). We replicate that here for full wire compatibility.

This module uses the pure-Python protobuf library (google.protobuf) to build
descriptors at runtime — no generated _pb2 files required.
"""

from __future__ import annotations

from google.protobuf import descriptor as _descriptor
from google.protobuf import descriptor_pool as _pool
from google.protobuf import symbol_database as _sym_db
from google.protobuf.internal import decoder as _decoder
from google.protobuf.internal import encoder as _encoder
from google.protobuf import message as _message
from google.protobuf import reflection as _reflection
from google.protobuf import descriptor_pb2

from . import models


# ─────────────────────────────────────────────────────────
#  Build the proto file descriptor at runtime
# ─────────────────────────────────────────────────────────

def _build_file_descriptor() -> descriptor_pb2.FileDescriptorProto:
    """Build a FileDescriptorProto matching the server's wire format."""
    f = descriptor_pb2.FileDescriptorProto()
    f.name = "keyvault.proto"
    f.package = "tau.keyvault"
    f.syntax = "proto3"

    # ── bcl_DateTime (protobuf-net internal) ──
    _add_message(f, "bcl_DateTime", [
        (1, "value", descriptor_pb2.FieldDescriptorProto.TYPE_SINT64),
        (2, "scale", descriptor_pb2.FieldDescriptorProto.TYPE_INT32),
    ])

    # ── Response messages ──

    _add_message(f, "KeyEntryResponse", [
        (1, "key", descriptor_pb2.FieldDescriptorProto.TYPE_STRING),
        (2, "value", descriptor_pb2.FieldDescriptorProto.TYPE_STRING),
        (3, "environment", descriptor_pb2.FieldDescriptorProto.TYPE_STRING),
        (4, "data_type", descriptor_pb2.FieldDescriptorProto.TYPE_STRING),
        (5, "is_sensitive", descriptor_pb2.FieldDescriptorProto.TYPE_BOOL),
        (6, "updated_at", descriptor_pb2.FieldDescriptorProto.TYPE_MESSAGE, "bcl_DateTime"),
    ])

    _add_message(f, "KeyEntryListResponse", [
        (1, "items", descriptor_pb2.FieldDescriptorProto.TYPE_MESSAGE, "KeyEntryResponse",
         descriptor_pb2.FieldDescriptorProto.LABEL_REPEATED),
    ])

    _add_message(f, "EnvironmentListResponse", [
        (1, "environments", descriptor_pb2.FieldDescriptorProto.TYPE_STRING, None,
         descriptor_pb2.FieldDescriptorProto.LABEL_REPEATED),
    ])

    _add_message(f, "DeleteEnvironmentResponse", [
        (1, "message", descriptor_pb2.FieldDescriptorProto.TYPE_STRING),
        (2, "deleted_keys", descriptor_pb2.FieldDescriptorProto.TYPE_INT32),
    ])

    _add_message(f, "RenameEnvironmentResponse", [
        (1, "message", descriptor_pb2.FieldDescriptorProto.TYPE_STRING),
        (2, "updated_keys", descriptor_pb2.FieldDescriptorProto.TYPE_INT32),
    ])

    _add_message(f, "ExportKeyItemResponse", [
        (1, "key", descriptor_pb2.FieldDescriptorProto.TYPE_STRING),
        (2, "value", descriptor_pb2.FieldDescriptorProto.TYPE_STRING),
        (3, "data_type", descriptor_pb2.FieldDescriptorProto.TYPE_STRING),
        (4, "is_sensitive", descriptor_pb2.FieldDescriptorProto.TYPE_BOOL),
    ])

    _add_message(f, "ExportPayloadResponse", [
        (1, "version", descriptor_pb2.FieldDescriptorProto.TYPE_STRING),
        (2, "export_date", descriptor_pb2.FieldDescriptorProto.TYPE_MESSAGE, "bcl_DateTime"),
        (3, "environment", descriptor_pb2.FieldDescriptorProto.TYPE_STRING),
        (4, "key_count", descriptor_pb2.FieldDescriptorProto.TYPE_INT32),
        (5, "keys", descriptor_pb2.FieldDescriptorProto.TYPE_MESSAGE, "ExportKeyItemResponse",
         descriptor_pb2.FieldDescriptorProto.LABEL_REPEATED),
    ])

    _add_message(f, "ImportResultResponse", [
        (1, "imported", descriptor_pb2.FieldDescriptorProto.TYPE_INT32),
        (2, "skipped", descriptor_pb2.FieldDescriptorProto.TYPE_INT32),
        (3, "message", descriptor_pb2.FieldDescriptorProto.TYPE_STRING),
    ])

    _add_message(f, "ErrorResponse", [
        (1, "error", descriptor_pb2.FieldDescriptorProto.TYPE_STRING),
    ])

    # ── Request messages ──

    _add_message(f, "UpsertRequest", [
        (1, "key", descriptor_pb2.FieldDescriptorProto.TYPE_STRING),
        (2, "value", descriptor_pb2.FieldDescriptorProto.TYPE_STRING),
        (3, "environment", descriptor_pb2.FieldDescriptorProto.TYPE_STRING),
        (4, "data_type", descriptor_pb2.FieldDescriptorProto.TYPE_STRING),
        (5, "is_sensitive", descriptor_pb2.FieldDescriptorProto.TYPE_BOOL),
    ])

    _add_message(f, "RenameRequest", [
        (1, "new_name", descriptor_pb2.FieldDescriptorProto.TYPE_STRING),
    ])

    _add_message(f, "ImportKeyItem", [
        (1, "key", descriptor_pb2.FieldDescriptorProto.TYPE_STRING),
        (2, "value", descriptor_pb2.FieldDescriptorProto.TYPE_STRING),
        (3, "data_type", descriptor_pb2.FieldDescriptorProto.TYPE_STRING),
        (4, "is_sensitive", descriptor_pb2.FieldDescriptorProto.TYPE_BOOL),
    ])

    _add_message(f, "ImportRequest", [
        (1, "environment", descriptor_pb2.FieldDescriptorProto.TYPE_STRING),
        (2, "mode", descriptor_pb2.FieldDescriptorProto.TYPE_STRING),
        (3, "keys", descriptor_pb2.FieldDescriptorProto.TYPE_MESSAGE, "ImportKeyItem",
         descriptor_pb2.FieldDescriptorProto.LABEL_REPEATED),
    ])

    return f


def _add_message(file_proto, name, fields):
    """Helper to add a message type to a FileDescriptorProto."""
    msg = file_proto.message_type.add()
    msg.name = name
    for field_def in fields:
        fld = msg.field.add()
        fld.number = field_def[0]
        fld.name = field_def[1]
        fld.type = field_def[2]
        fld.label = descriptor_pb2.FieldDescriptorProto.LABEL_OPTIONAL
        if len(field_def) > 3 and field_def[3] is not None:
            fld.type_name = f".tau.keyvault.{field_def[3]}"
        if len(field_def) > 4:
            fld.label = field_def[4]


# ─────────────────────────────────────────────────────────
#  Register with the pool and build message classes
# ─────────────────────────────────────────────────────────

_file_proto = _build_file_descriptor()
_pool_inst = _pool.Default()
_pool_inst.Add(_file_proto)

_sym = _sym_db.Default()


def _make_class(full_name: str):
    """Create a protobuf message class from a registered descriptor."""
    desc = _pool_inst.FindMessageTypeByName(full_name)
    cls = _reflection.GeneratedProtocolMessageType(
        desc.name,
        (_message.Message,),
        {"DESCRIPTOR": desc, "__module__": __name__},
    )
    _sym.RegisterMessage(cls)
    return cls


_BclDateTime = _make_class("tau.keyvault.bcl_DateTime")

_PbKeyEntryResponse = _make_class("tau.keyvault.KeyEntryResponse")
_PbKeyEntryListResponse = _make_class("tau.keyvault.KeyEntryListResponse")
_PbEnvironmentListResponse = _make_class("tau.keyvault.EnvironmentListResponse")
_PbDeleteEnvironmentResponse = _make_class("tau.keyvault.DeleteEnvironmentResponse")
_PbRenameEnvironmentResponse = _make_class("tau.keyvault.RenameEnvironmentResponse")
_PbExportKeyItemResponse = _make_class("tau.keyvault.ExportKeyItemResponse")
_PbExportPayloadResponse = _make_class("tau.keyvault.ExportPayloadResponse")
_PbImportResultResponse = _make_class("tau.keyvault.ImportResultResponse")
_PbErrorResponse = _make_class("tau.keyvault.ErrorResponse")

_PbUpsertRequest = _make_class("tau.keyvault.UpsertRequest")
_PbRenameRequest = _make_class("tau.keyvault.RenameRequest")
_PbImportKeyItem = _make_class("tau.keyvault.ImportKeyItem")
_PbImportRequest = _make_class("tau.keyvault.ImportRequest")


# ═══════════════════════════════════════════════════════════
#  Public encode / decode helpers
# ═══════════════════════════════════════════════════════════

# Maps model type → (protobuf class, decode function)


def _decode_key_entry(pb) -> models.KeyEntryResponse:
    return models.KeyEntryResponse(
        key=pb.key,
        value=pb.value,
        environment=pb.environment,
        data_type=pb.data_type,
        is_sensitive=pb.is_sensitive,
        updated_at=None,  # bcl.DateTime is opaque; string is preferred via JSON
    )


def decode_key_entry_response(data: bytes) -> models.KeyEntryResponse:
    """Decode a KeyEntryResponse from protobuf bytes."""
    pb = _PbKeyEntryResponse()
    pb.ParseFromString(data)
    return _decode_key_entry(pb)


def decode_key_entry_list_response(data: bytes) -> models.KeyEntryListResponse:
    """Decode a KeyEntryListResponse from protobuf bytes."""
    pb = _PbKeyEntryListResponse()
    pb.ParseFromString(data)
    return models.KeyEntryListResponse(
        items=[_decode_key_entry(item) for item in pb.items],
    )


def decode_environment_list_response(data: bytes) -> models.EnvironmentListResponse:
    pb = _PbEnvironmentListResponse()
    pb.ParseFromString(data)
    return models.EnvironmentListResponse(environments=list(pb.environments))


def decode_delete_environment_response(data: bytes) -> models.DeleteEnvironmentResponse:
    pb = _PbDeleteEnvironmentResponse()
    pb.ParseFromString(data)
    return models.DeleteEnvironmentResponse(message=pb.message, deleted_keys=pb.deleted_keys)


def decode_rename_environment_response(data: bytes) -> models.RenameEnvironmentResponse:
    pb = _PbRenameEnvironmentResponse()
    pb.ParseFromString(data)
    return models.RenameEnvironmentResponse(message=pb.message, updated_keys=pb.updated_keys)


def _decode_export_key_item(pb) -> models.ExportKeyItemResponse:
    return models.ExportKeyItemResponse(
        key=pb.key, value=pb.value,
        data_type=pb.data_type, is_sensitive=pb.is_sensitive,
    )


def decode_export_payload_response(data: bytes) -> models.ExportPayloadResponse:
    pb = _PbExportPayloadResponse()
    pb.ParseFromString(data)
    return models.ExportPayloadResponse(
        version=pb.version,
        export_date=None,
        environment=pb.environment,
        key_count=pb.key_count,
        keys=[_decode_export_key_item(k) for k in pb.keys],
    )


def decode_import_result_response(data: bytes) -> models.ImportResultResponse:
    pb = _PbImportResultResponse()
    pb.ParseFromString(data)
    return models.ImportResultResponse(
        imported=pb.imported, skipped=pb.skipped, message=pb.message,
    )


def decode_error_response(data: bytes) -> str | None:
    """Decode an ErrorResponse and return the error string, or None."""
    try:
        pb = _PbErrorResponse()
        pb.ParseFromString(data)
        return pb.error or None
    except Exception:
        return None


# ── Encode helpers ──

def encode_upsert_request(
    key: str, value: str, environment: str,
    data_type: str, is_sensitive: bool,
) -> bytes:
    pb = _PbUpsertRequest()
    pb.key = key
    pb.value = value
    pb.environment = environment
    pb.data_type = data_type
    pb.is_sensitive = is_sensitive
    return pb.SerializeToString()


def encode_rename_request(new_name: str) -> bytes:
    pb = _PbRenameRequest()
    pb.new_name = new_name
    return pb.SerializeToString()


def encode_import_request(
    environment: str | None,
    mode: str,
    keys: list[models.ImportKeyItem],
) -> bytes:
    pb = _PbImportRequest()
    if environment:
        pb.environment = environment
    pb.mode = mode
    for k in keys:
        item = pb.keys.add()
        item.key = k.key
        item.value = k.value
        item.data_type = k.data_type
        item.is_sensitive = k.is_sensitive
    return pb.SerializeToString()
